security/s2a/internal/crypter/expander.go

```diff
@@ -12,3 +14,24 @@ type hkdfExpander interface {
 	// in number of bytes, and returns the resulting expanded key.
 	expand(hash func() hash.Hash, secret, label []byte, length int) ([]byte, error)
 }
+
+// defaultHKDFExpander is the default HKDF expander which uses Go's crypto/hkdf
+// for HKDF expansion.
+type defaultHKDFExpander struct{}
+
+// newDefaultHKDFExpander creates an instance of the default HKDF expander.
+func newDefaultHKDFExpander() hkdfExpander {
+	return &defaultHKDFExpander{}
+}
+
+func (*defaultHKDFExpander) expand(hash func() hash.Hash, secret, label []byte, length int) ([]byte, error) {
+	outBuf := make([]byte, length)
+	n, err := hkdf.Expand(hash, secret, label).Read(outBuf)
+	if err != nil {
+		return nil, fmt.Errorf("hkdf.expand.Read failed with error: %v", err)
+	}
+	if n != length {
+		return nil, fmt.Errorf("hkdf.expand.Read returned unexpected length, got: %d, want: %d", n, length)
```

Comment:
To me it seems that this must not happen. If the caller provided the right `hash` with the right `length`, then `n` must be equal to `length`. If that's the case, then maybe we should `panic` here. However, we might not need this check at all. See my comment above.

Reply:
Looking through the implementation of `Read` for an `hdkf` object, I don't think it's possible to ever return an invalid length while not returning an error. Since we check for an error in the line above, I don't think its possible to hit this condition.

However, I think the documentation of the `Read` interface suggests that other implementations may return a nil error while still returning a number that is less than the number that was requested for. As you mentioned though, if we provide the right `hash` with the right `length`, I think hitting this condition may signal a programming error and so `panic` may be the correct here.

Reply:
Since `length` is no longer a parameter, I don't think we should `panic` here. I left it as an error for now, I don't know if it's worth doing the check at all however.